### ISSUE TITLE
EM simulation: Add checks before Pardiso runs

### ISSUE THREAD

**Issue by thibaut-kobold** [simpeg/simpeg#1071]
The problem:
Right now, when launching an EM simulation, Pardiso runs first before making the two following checks:
- Are all the receivers within the domain?
- Is the requested field for the receivers available for the chosen simulation? (not all combinations of sources, simulations, and receivers are possible at the moment).

In consequence, for decent size modelling, you can wait > 20 minutes before getting an error that informs you the requested field is not available.

The fix:
- the two points above can be tested before the Pardiso computation happens.

**Comment by ckohnke** [comment]
Yes! Both of these were major pain points for me when starting out with SimPEG, and I still sometimes forget about the simulation part. Thoughts to start off implementation below -- 

Checking receivers in domain should be straightforward. The only sticky implementation would probably be cylindrical meshes (implementation allows for receiver definitions in 3D while the mesh is 2D), but that's not too different from the Tensor/Tree cases. Or any mesh that is defined in a low dimension (e.g. 1D mesh for layered earth) for the simulation, but the returned fields are valid for a higher dimension (3D receiver locations). Should probably also check the source location for methods with physical source locations, but it's the same logic.

    def check_receivers_in_mesh(mesh, rx_list):
        urx = list of unique rx locations
        for rx in urx:
            # check if rx in mesh
            # raise error / return False if not
        return True

The simulation one I think is more involved. It's trivial if the simulation knows what fields it computes, and what fields will be needed at the receivers, but I don't think that's the case. Fields objects know what fields they hold, and what fields they can compute, but I don't think the Fields inherently know about the receivers. It's only when the Fields go through the projection matrix to compute data at the receivers that the error occurs because there's no way to project from the Fields object to what the receivers want. A dirty way of doing it may be to somehow create an empty Fields object from the simulation and then compare the known/aliased fields with the receiver type. Might have issues checking receivers that are a combination of different field types (MT Impedance comes to mind).

    def check_simulation_fields(simulation, rx_list):
        urx = list of unique rx types
        sim_fields = simulation.make_null_fields() # some method to get an empty Fields object appropriate to the simulation
        for rx in urx:
            # check if rx in sim_fields
            # raise error / return False if not
        return True

Just making sure, is this unique to EM in SimPEG? Potential fields probably does not have this issue, but I haven't looked. Seismic would definitely have this same issue, and I'd expect the flow module too. So these checks should probably be placed in BaseSimulation (or maybe Base[X]Simulation for each method that needs them)? And then actually run when the simulation calls fields() and/or dpred(). Maybe when the simulation is initialized instead, but then there's issues with using `simulation.survey=survey`... The source/receiver in mesh check may be best handled mainly in Discretize (it is just a point in mesh problem, would just need to get locations to check, and I think the method exists already) and then called when needed.
